src/NuGet.Core/NuGet.PackageManagement/Resolution/PrunePackageTree.cs

```diff
@@ -63,30 +63,32 @@ private static void CollectAllowedFromDependenciesOfPackagesToInstall(IEnumerabl
         {
             var prereleasePackageToInstall = new HashSet<PackageIdentity>(packagesToInstall.Where(p => p.HasVersion && p.Version.IsPrerelease), PackageIdentityComparer.Default);
 
-            var maxDepth = packages.Select(p => p.Id).Distinct(StringComparer.OrdinalIgnoreCase).Count();
+            var visitedNodes = new HashSet<SourcePackageDependencyInfo>();
 
             foreach (var packageToInstall in packages.Where(p => prereleasePackageToInstall.Contains(p)))
             {
-                int depth = 0;
-                WalkDependencies(packages, packageToInstall, allowed, depth, maxDepth);
+                WalkDependencies(packages, packageToInstall, allowed, visitedNodes);
             }
         }
 
-        private static void WalkDependencies(IEnumerable<SourcePackageDependencyInfo> packages, SourcePackageDependencyInfo packageToInstall, HashSet<string> allowed, int depth, int maxDepth)
+        private static void WalkDependencies(IEnumerable<SourcePackageDependencyInfo> packages, SourcePackageDependencyInfo packageToInstall, HashSet<string> allowed, HashSet<SourcePackageDependencyInfo> visitedNodes)
```

Comment:
yeah it will save further time to execute linq expression every time for a similar package id. I'll better use ILookup interface